pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,552 @@
+		return nil, errs.ErrReservedGCBarrierID.GenWithStackByArgs(barrierID)
+	}
+	// Disallow empty barrierID
+	if len(barrierID) == 0 {
+		return nil, errs.ErrInvalidArgument.GenWithStackByArgs("barrierID", barrierID)
+	}
+
+	var expirationTime *time.Time = nil
+	if ttl < time.Duration(math.MaxInt64) {
+		t := now.Add(ttl)
+		expirationTime = &t
+	}
+	newBarrier := endpoint.NewGCBarrier(barrierID, barrierTS, expirationTime)
+
+	err := m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		txnSafePoint, err1 := m.gcMetaStorage.LoadTxnSafePoint(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+		if barrierTS < txnSafePoint {
+			return errs.ErrGCBarrierTSBehindTxnSafePoint.GenWithStackByArgs(barrierTS, txnSafePoint)
+		}
+		err1 = wb.SetGCBarrier(keyspaceID, newBarrier)
+		return err1
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newBarrier, nil
+}
+
+// DeleteGCBarrier deletes a GC barrier by the given barrierID. Returns the information of the deleted GC barrier, or
+// nil if the barrier does not exist.
+//
+// When this method is called on a keyspace without keyspace-level GC enabled, it will be equivalent to calling it on
+// the NullKeyspace.
+func (m *GCStateManager) DeleteGCBarrier(keyspaceID uint32, barrierID string) (*endpoint.GCBarrier, error) {
+	keyspaceID, err := m.redirectKeyspace(keyspaceID, true)
+	if err != nil {
+		return nil, err
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.deleteGCBarrierImpl(keyspaceID, barrierID)
+}
+
+func (m *GCStateManager) deleteGCBarrierImpl(keyspaceID uint32, barrierID string) (*endpoint.GCBarrier, error) {
+	// The barrier ID (or service ID of the service safe points) is reserved for keeping backward compatibility.
+	if keyspaceID == constant.NullKeyspaceID && barrierID == keypath.GCWorkerServiceSafePointID {
+		return nil, errs.ErrReservedGCBarrierID.GenWithStackByArgs(barrierID)
+	}
+	// Disallow empty barrierID
+	if len(barrierID) == 0 {
+		return nil, errs.ErrInvalidArgument.GenWithStackByArgs("barrierID", barrierID)
+	}
+
+	var deletedBarrier *endpoint.GCBarrier
+	err := m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		var err1 error
+		deletedBarrier, err1 = m.gcMetaStorage.LoadGCBarrier(keyspaceID, barrierID)
+		if err1 != nil {
+			return err1
+		}
+		return wb.DeleteGCBarrier(keyspaceID, barrierID)
+	})
+	return deletedBarrier, err
+}
+
+// getGCStateInTransaction gets all properties in GC states within a context of gcMetaStorage.RunInGCStateTransaction.
+// This read only and won't write anything to the GCStateWriteBatch. It still receives a write batch to ensure
+// it's running in a transactional context.
+func (m *GCStateManager) getGCStateInTransaction(keyspaceID uint32, _ *endpoint.GCStateWriteBatch) (GCState, error) {
+	result := GCState{
+		KeyspaceID: keyspaceID,
+	}
+	if keyspaceID != constant.NullKeyspaceID {
```

Comment:
"In this code, why do we set result.IsKeyspaceLevel = true when checking if keyspaceID != constant.NullKeyspaceID?"

Reply:
If it's not keyspace level GC, the call to this function must have been redirected to the null keyspace. I'll add comments to explain this later.